feat: support sm2::dsa::SigningKey for x509_cert::builder::RequestBuilder::new

version:

```toml
sm2 = "0.13.3"
```

example:

```rust
pub fn generate_sm2_csr(pri_key: SecretKey, name: Name) {
    let signer = sm2::dsa::SigningKey::new("", &pri_key).unwrap();
    let csr = RequestBuilder::new(name, &signer)
      .and_then(|b| b.build()).unwrap();
  }
```

error message:

```txt
error[E0277]: the trait bound `sm2::dsa::VerifyingKey: EncodePublicKey` is not satisfied
   --> src/csr.rs:32:41
    |
32  |     let csr = RequestBuilder::new(name, &signer)
    |               -------------------       ^^^^^^^ the trait `EncodePublicKey` is not implemented for `sm2::dsa::VerifyingKey`
    |               |
    |               required by a bound introduced by this call
    |
    = help: the following other types implement trait `EncodePublicKey`:
              rsa::RsaPublicKey
              rsa::pkcs1v15::VerifyingKey<D>
              rsa::pss::VerifyingKey<D>
              sm2::elliptic_curve::PublicKey<C>
note: required by a bound in `RequestBuilder::<'s, S>::new`
```

it seems some impl is missing, like `impl SignatureBitStringEncoding for Signature`, `impl SignatureAlgorithmIdentifier for VerifyingKey`, and so on. Crypto code is challenging to me, and I fail to reair it myself.

[GMObjectIdentifiers](https://github.com/bcgit/bc-java/blob/main/core/src/main/java/org/bouncycastle/asn1/gm/GMObjectIdentifiers.java) .

--- a/sm2/CHANGELOG.md
+++ b/sm2/CHANGELOG.md
@@ -5,6 +5,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## 0.14.0 (UNRELEASED)
+## Added 
+- Signature and key PKCS8 support ([#1127])
+
 ## Changed
 - Update to `elliptic-curve` v0.14 ([#1011])
 - Update to `secdect` v0.3 ([#1084])
@@ -16,7 +19,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 [#1011]: https://github.com/RustCrypto/elliptic-curves/pull/1011
 [#1084]: https://github.com/RustCrypto/elliptic-curves/pull/1084
 [#1125]: https://github.com/RustCrypto/elliptic-curves/pull/1125
-
+[#1127]: https://github.com/RustCrypto/elliptic-curves/pull/1127
 
 ## 0.13.3 (2023-11-20)
 ### Added

--- a/sm2/src/dsa.rs
+++ b/sm2/src/dsa.rs
@@ -50,6 +50,13 @@ use signature::{Error, Result, SignatureEncoding};
 #[cfg(feature = "alloc")]
 use alloc::vec::Vec;
 
+#[cfg(feature = "pkcs8")]
+use crate::pkcs8::{
+    AlgorithmIdentifierRef, ObjectIdentifier, der::AnyRef, spki::AssociatedAlgorithmIdentifier,
+};
+#[cfg(all(feature = "alloc", feature = "pkcs8"))]
+use crate::pkcs8::{der, spki::SignatureBitStringEncoding};
+
 /// SM2DSA signature serialized as bytes.
 pub type SignatureBytes = [u8; Signature::BYTE_SIZE];
 
@@ -196,4 +203,30 @@ impl TryFrom<&[u8]> for Signature {
     fn try_from(bytes: &[u8]) -> Result<Signature> {
         Signature::from_slice(bytes)
     }
+}
+
+#[cfg(all(feature = "alloc", feature = "pkcs8"))]
+impl SignatureBitStringEncoding for Signature {
+    fn to_bitstring(&self) -> der::Result<der::asn1::BitString> {
+        der::asn1::BitString::new(0, self.to_vec())
+    }
+}
+
+#[cfg(all(feature = "alloc", feature = "pkcs8"))]
+impl AssociatedAlgorithmIdentifier for Signature {
+    type Params = AnyRef<'static>;
+
+    const ALGORITHM_IDENTIFIER: AlgorithmIdentifierRef<'static> = AlgorithmIdentifierRef {
+        // Reference: https://datatracker.ietf.org/doc/html/draft-sca-cfrg-sm3#section-8.1.3
+        // ```
+        // "1.2.156.10197.1.501" for "Digital Signature: SM2 and SM3"
+        // o  "id-dsa-sm2sm3" "{id-int dsa-sm2sm3(501)}"
+        // ```
+        //
+        // NOTE: Here [`Signature`] represent an SM2 signed object with an SM3 hash (implementation detail
+        // of [`sm2::dsa::SigningKey`] and [`sm2::dsa::VerifyingKey`], that might need to change
+        // once/if the hash function gets modular.
+        oid: ObjectIdentifier::new_unwrap("1.2.156.10197.1.501"),
+        parameters: None,
+    };
 }

--- a/sm2/src/dsa/signing.rs
+++ b/sm2/src/dsa/signing.rs
@@ -33,6 +33,12 @@ use signature::{
 };
 use sm3::Sm3;
 
+#[cfg(feature = "pkcs8")]
+use crate::pkcs8::{
+    der::AnyRef,
+    spki::{AlgorithmIdentifier, AssociatedAlgorithmIdentifier, SignatureAlgorithmIdentifier},
+};
+
 /// SM2DSA secret key used for signing messages and producing signatures.
 ///
 /// ## Usage
@@ -222,4 +228,12 @@ fn sign_prehash_rfc6979(secret_scalar: &Scalar, prehash: &[u8], data: &[u8]) -> 
 
     // A7: the digital signature of M is (r, s)
     Signature::from_scalars(r, s)
+}
+
+#[cfg(feature = "pkcs8")]
+impl SignatureAlgorithmIdentifier for SigningKey {
+    type Params = AnyRef<'static>;
+
+    const SIGNATURE_ALGORITHM_IDENTIFIER: AlgorithmIdentifier<Self::Params> =
+        Signature::ALGORITHM_IDENTIFIER;
 }

--- a/sm2/src/dsa/verifying.rs
+++ b/sm2/src/dsa/verifying.rs
@@ -30,6 +30,14 @@ use sm3::{Sm3, digest::Digest};
 #[cfg(feature = "alloc")]
 use alloc::{boxed::Box, string::String};
 
+#[cfg(all(feature = "alloc", feature = "pkcs8"))]
+use crate::pkcs8::{self, EncodePublicKey, spki};
+#[cfg(feature = "pkcs8")]
+use crate::pkcs8::{
+    der::AnyRef,
+    spki::{AlgorithmIdentifier, AssociatedAlgorithmIdentifier, SignatureAlgorithmIdentifier},
+};
+
 /// SM2DSA public key used for verifying signatures are valid for a given
 /// message.
 ///
@@ -200,4 +208,19 @@ impl ToEncodedPoint<Sm2> for VerifyingKey {
     fn to_encoded_point(&self, compress: bool) -> EncodedPoint {
         self.as_affine().to_encoded_point(compress)
     }
+}
+
+#[cfg(all(feature = "alloc", feature = "pkcs8"))]
+impl EncodePublicKey for VerifyingKey {
+    fn to_public_key_der(&self) -> spki::Result<pkcs8::Document> {
+        self.public_key.to_public_key_der()
+    }
+}
+
+#[cfg(feature = "pkcs8")]
+impl SignatureAlgorithmIdentifier for VerifyingKey {
+    type Params = AnyRef<'static>;
+
+    const SIGNATURE_ALGORITHM_IDENTIFIER: AlgorithmIdentifier<Self::Params> =
+        Signature::ALGORITHM_IDENTIFIER;
 }

--- a/sm2/tests/pkcs8.rs
+++ b/sm2/tests/pkcs8.rs
@@ -14,6 +14,15 @@ use {
     sm2::elliptic_curve::pkcs8::{EncodePrivateKey, EncodePublicKey},
 };
 
+#[cfg(all(feature = "alloc", feature = "pkcs8"))]
+use {
+    signature::{Keypair, Signer},
+    sm2::{
+        dsa::{Signature, SigningKey},
+        pkcs8::spki::{DynSignatureAlgorithmIdentifier, SignatureBitStringEncoding},
+    },
+};
+
 /// DER-encoded PKCS#8 private key
 const PKCS8_PRIVATE_KEY_DER: &[u8; 138] = include_bytes!("examples/pkcs8-private-key.der");
 
@@ -122,4 +131,22 @@ fn encode_pkcs8_public_key_to_pem() {
         sm2::PublicKey::from_public_key_der(&PKCS8_PUBLIC_KEY_DER[..]).unwrap();
     let reencoded_public_key = original_public_key.to_string();
     assert_eq!(reencoded_public_key.as_str(), PKCS8_PUBLIC_KEY_PEM);
+}
+
+#[test]
+fn test_x509() {
+    fn dummy_cert_builder<S, Signature>(_signer: &S)
+    where
+        S: Signer<Signature>,
+        S: Keypair + DynSignatureAlgorithmIdentifier,
+        S::VerifyingKey: EncodePublicKey,
+        Signature: SignatureBitStringEncoding,
+    {
+        // we just want to check the trait bounds here
+    }
+
+    let secret_key = PKCS8_PRIVATE_KEY_PEM.parse::<sm2::SecretKey>().unwrap();
+    const IDENTITY: &str = "example@rustcrypto.org";
+    let signing_key = SigningKey::new(IDENTITY, &secret_key).unwrap();
+    let _signature = dummy_cert_builder::<_, Signature>(&signing_key);
 }

